File: app/android-widget/java/com/jhona/organizacao/widget/RoutineWidgetProvider.java
package com.jhona.organizacao.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.widget.RemoteViews;

import com.jhona.organizacao.R;
import org.json.JSONArray;
import org.json.JSONObject;

public class RoutineWidgetProvider extends AppWidgetProvider {

    // Nome do SharedPreferences usado pelo @capacitor/preferences
    private static final String PREFS_NAME = "CapacitorPreferences";
    private static final String PREFS_KEY = "widget_data";
    private static final int MAX_ITEMS = 5;

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        android.content.ComponentName cn =
            new android.content.ComponentName(context, RoutineWidgetProvider.class);
        int[] ids = manager.getAppWidgetIds(cn);
        onUpdate(context, manager, ids);
    }

    private void updateWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {
        try {
            RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_layout_minimal);

            // Lê dados do SharedPreferences (escritos pelo app via @capacitor/preferences)
            SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            String jsonString = prefs.getString(PREFS_KEY, "");

            if (jsonString.isEmpty()) {
                views.setTextViewText(R.id.widget_text, "📋 Minha Vida\nAbra o app para carregar");
            } else {
                try {
                    JSONObject data = new JSONObject(jsonString);
                    JSONArray items = data.optJSONArray("items");
                    String progress = data.optString("progress", "0%");
                    String dateLabel = data.optString("date", "");

                    StringBuilder sb = new StringBuilder();
                    sb.append("📋 ").append(progress).append(" ").append(dateLabel).append("\n");

                    if (items != null && items.length() > 0) {
                        int count = Math.min(items.length(), MAX_ITEMS);
                        for (int i = 0; i < count; i++) {
                            JSONObject item = items.getJSONObject(i);
                            boolean done = item.optBoolean("done", false);
                            String emoji = item.optString("emoji", "");
                            String activity = item.optString("activity", "");
                            String time = item.optString("time", "");
                            sb.append(done ? "✅" : "⏳")
                              .append(" ").append(time)
                              .append(" ").append(emoji).append(" ").append(activity)
                              .append("\n");
                        }
                    }

                    views.setTextViewText(R.id.widget_text, sb.toString().trim());

                } catch (Exception e) {
                    views.setTextViewText(R.id.widget_text, "📋 Minha Vida\nDados carregados ✅");
                }
            }

            // Click: abre o app
            Intent intent = context.getPackageManager().getLaunchIntentForPackage(context.getPackageName());
            if (intent != null) {
                PendingIntent pi = PendingIntent.getActivity(
                    context, 0, intent,
                    PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
                );
                views.setOnClickPendingIntent(R.id.widget_text, pi);
            }

            appWidgetManager.updateAppWidget(appWidgetId, views);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/android-widget/java/com/jhona/organizacao/widget/RoutineWidgetProvider.kt
package com.jhona.organizacao.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import org.json.JSONArray
import org.json.JSONObject

class RoutineWidgetProvider : AppWidgetProvider() {

    companion object {
        private val ITEM_VIEWS = intArrayOf(
            R.id.item_1, R.id.item_2, R.id.item_3, R.id.item_4, R.id.item_5
        )
        private val ITEM_CHECKS = intArrayOf(
            R.id.item_1_check, R.id.item_2_check, R.id.item_3_check,
            R.id.item_4_check, R.id.item_5_check
        )
        private val ITEM_TIMES = intArrayOf(
            R.id.item_1_time, R.id.item_2_time, R.id.item_3_time,
            R.id.item_4_time, R.id.item_5_time
        )
        private val ITEM_TEXTS = intArrayOf(
            R.id.item_1_text, R.id.item_2_text, R.id.item_3_text,
            R.id.item_4_text, R.id.item_5_text
        )
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray
    ) {
        for (appWidgetId in appWidgetIds) {
            updateWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
        // Called when the first widget instance is added — force immediate update
        val manager = AppWidgetManager.getInstance(context)
        val componentName = android.content.ComponentName(
            context, RoutineWidgetProvider::class.java
        )
        val ids = manager.getAppWidgetIds(componentName)
        onUpdate(context, manager, ids)
    }

    private fun updateWidget(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetId: Int
    ) {
        val views = RemoteViews(context.packageName, R.layout.widget_layout)

        // --- Read data from SharedPreferences (written by the web app) ---
        val prefs = context.getSharedPreferences("MinhaVidaWidget", Context.MODE_PRIVATE)
        val jsonString = prefs.getString("widget_data", "{}") ?: "{}"

        try {
            val data = JSONObject(jsonString)
            val items = data.optJSONArray("items")
            val progress = data.optString("progress", "0%")
            val dateLabel = data.optString("date", "")

            // Update progress header
            views.setTextViewText(R.id.widget_progress, "$progress $dateLabel")

            if (items != null && items.length() > 0) {
                views.setViewVisibility(R.id.widget_empty, android.view.View.GONE)

                val count = minOf(items.length(), 5)
                for (i in 0 until count) {
                    val item = items.getJSONObject(i)
                    views.setViewVisibility(ITEM_VIEWS[i], android.view.View.VISIBLE)

                    val isDone = item.optBoolean("done", false)
                    views.setTextViewText(ITEM_CHECKS[i], if (isDone) "✅" else "⏳")
                    views.setTextViewText(ITEM_TIMES[i], item.optString("time", ""))

                    val emoji = item.optString("emoji", "")
                    val activity = item.optString("activity", "")
                    views.setTextViewText(ITEM_TEXTS[i], "$emoji $activity")
                }

                // Hide remaining unused slots
                for (i in count until 5) {
                    views.setViewVisibility(ITEM_VIEWS[i], android.view.View.GONE)
                }
            } else {
                views.setViewVisibility(R.id.widget_empty, android.view.View.VISIBLE)
                for (i in 0 until 5) {
                    views.setViewVisibility(ITEM_VIEWS[i], android.view.View.GONE)
                }
            }
        } catch (e: Exception) {
            views.setViewVisibility(R.id.widget_empty, android.view.View.VISIBLE)
            views.setTextViewText(R.id.widget_empty, "📋 Abra o app para carregar")
            for (i in 0 until 5) {
                views.setViewVisibility(ITEM_VIEWS[i], android.view.View.GONE)
            }
        }

        // --- Click handler: open the app ---
        val intent = context.packageManager.getLaunchIntentForPackage(context.packageName)
        if (intent != null) {
            val pendingIntent = PendingIntent.getActivity(
                context, 0, intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            views.setOnClickPendingIntent(R.id.widget_footer, pendingIntent)
        }

        appWidgetManager.updateAppWidget(appWidgetId, views)
    }
}
